cr/state/state.go

```diff
@@ -280,15 +280,29 @@ func (s *State) registerCR(tx *types.Transaction, height uint32) {
 	}
 	candidate.depositAmount = amount
 
-	s.history.Append(height, func() {
-		s.Nicknames[nickname] = struct{}{}
-		s.CodeDIDMap[code] = info.DID
-		s.PendingCandidates[info.DID] = &candidate
-	}, func() {
-		delete(s.Nicknames, nickname)
-		delete(s.CodeDIDMap, code)
-		delete(s.PendingCandidates, info.DID)
-	})
+	c := s.getCandidateByDID(info.DID)
+	if c == nil {
+		s.history.Append(height, func() {
+			s.Nicknames[nickname] = struct{}{}
+			s.CodeDIDMap[code] = info.DID
+			s.PendingCandidates[info.DID] = &candidate
+		}, func() {
+			delete(s.Nicknames, nickname)
+			delete(s.CodeDIDMap, code)
+			delete(s.PendingCandidates, info.DID)
+		})
+	} else {
+		s.history.Append(height, func() {
+			delete(s.CanceledCandidates, c.Info().DID)
+			s.Nicknames[nickname] = struct{}{}
+			s.PendingCandidates[info.DID] = &candidate
+		}, func() {
+			delete(s.PendingCandidates, info.DID)
+			s.Nicknames[c.Info().NickName] = struct{}{}
```

Comment:
need to  delete(s.Nicknames, info.NickName)

Reply:
yes, this should be a mistake

Reply:
Fixed